### PR TITLE
mrc-4723 Check user org membership on browser login, and security refactor

### PR DESCRIPTION
**Pull request by EmmaLRussell**
This branch implements three separate tickets which, after some investigation, it turned out to make sense to implement together. These are:
- mrc-4702: Check user org membership on browser login (not just API login)
- mrc-4682: Use third party Github API package (kohsuke)
- mrc-4723: Refactor auth to re-use more code between API and browser paths

Essentially, I did an initial implementation of mrc-4702 which used built-in Spring Security OAuth2 classes to access the Github API using the token obtained from the OAuth2 redirected browser login. However when I looked into re-using this approach for Packit API login, this proved infeasible as the Spring Security machinery is quite intractable to make a partial use of (providing the PAT token via our own API rather than obtained during the OAuth2 handshake). So instead, I've refactored so that both paths now call the same client code, which makes use of the third party Github API package, replacing the previous 'roll your own' client. 

Key changes:
- `GithubClient` reimplemented as `GithubUserClient`, largely replicating code from OrderlyWeb. Authenticates with given token, and returns a user principal as well as checking org membership (checking team membership to follow in future ticket). 
- `GithubUserLoginService` renamed to `GithubAPILoginService` - this uses the new `GithubUserClient` and is called from `LoginController` to execute the API login path. 
- `OAuth2UserService` updated to also use `GithubUserClient` to check org membership
- `OAuth2FailureHandler` added, and included in `WebSecurityConfig`. This deals with exceptions thrown during browser login (including org membership check failure). Like the success handler, this redirects to `/redirect` in the front end, but passes an `error` parameter rather than `token`. 
- `BrowserRedirect` pulled out to share some redirect code between the success handler and failure handler.
- The failure handler also uses `PackitExceptionHandler` to deal with PackitAuthenticationExceptions - the handler looks up error messages from the resource bundle. The API login path is already using the exception handler. We have a new `PackitAuthenticationException` type as the failure handler only deals with exceptions inherited from `AuthenticationException`.
- Front end changes to allow the `Redirect` component to deal with redirections with error parameter rather than token parameter. In this case, commit the error and redirect to the login page, which will display the error. 

To test:
Make yourself an OAuth app on Github and set the client id and secret in `/api/app/src/main/resources/application.properties`
Success case: Make sure auth.githubAPIOrgs in  `/api/app/src/main/resources/config.properties` includes an org you are a member of before you run the app
Failure case: Make sure auth.githubAPIOrgs does not include an org you are a member of

Browser login: Run up the app and login from htpp://localhost:3000. You should see a sensible error message in the failure case.
API login:  Use your Github PAT in a post like `curl -v -X POST -H "Content-Type: application/json" -d '{"githubtoken": "[YOUR GITHUB PAT]"}' http://localhost:8080/auth/login/github`.  The PAT needs to have read:org permission. You should get a JWT token and 200 response back in success case, or an error with 401 response in failure case.

### CODE REVIEW

File: api/app/src/main/kotlin/packit/clients/GithubUserClient.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        checkAuthenticated()
+        val ghu = ghUser!!
+
+        val user = User(
+            1L,
+            ghu.login,
+            "",
+            Role.USER,
+            ghu.name ?: "",
+            mutableMapOf()
```
Comment (EmmaLRussell) [inline api/app/src/main/kotlin/packit/clients/GithubUserClient.kt:36]: We're going to simplify the user class in a future ticket - should just need a string id from the provider, not a numeric id or any further metadata like name.

File: api/app/src/main/kotlin/packit/clients/GithubUserClient.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        )
+        return UserPrincipal.create(user, mutableMapOf())
+    }
+
+    fun checkGithubOrgMembership()
+    {
+        checkAuthenticated()
+
+        val userOrgs = ghUser!!.allOrganizations.map {org -> org.login}
+        val inAllowedOrg = allowedOrgs.any {allowed -> userOrgs.contains(allowed)}
```
Comment (EmmaLRussell) [inline api/app/src/main/kotlin/packit/clients/GithubUserClient.kt:46]: Currently supporting membership of multiple orgs, but as discussed, will replace this in next ticket with single allowed org, with optional team membership

**Review by absternator — COMMENTED** [review]
LGTM and works like a charm... Just want to double check what org permissions are needed... eg. i had mrc-ide but have to request approval to be able to use it....i only could use reside as i am an owner

**Review by M-Kusumgar — COMMENTED** [review]
Looks really good, was easy to understand so thank you for that! couple of things:

Are we going to handle the case where someone has the wrong secret for the id? Because currently you get this:

image
(we may want to sort out the styling at some point in the future too)

Are we also going to handle the wrong client id? because right now we go to a 404 page on github, wonder if we can just see the response from github and then tell the user their id does not exist? (these are probably suggestions for other tickets unless any of them are unintended)

One last weird thing (but this might not be relevant, could be to do with how the app is brought up) if you:
1. login with github successfully
2. kill the docker containers running on your machine
3. change the id in application properties
4. re build everything
5. on load you are immediately logged in and can do everything even though your client id is currently incorrect

Not sure if this is just because I have done something weird but perhaps we are not guarding certain routes correctly on load?

File: api/app/src/main/kotlin/packit/security/oauth2/OAuth2FailureHandler.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    )
+    {
+        val message = if (exception is PackitAuthenticationException) {
+            exceptionHandler.errorDetailForPackitAuthenticationException(exception).detail
+        }
+        else {
+            exception.message
+        }
+
+        val qs = LinkedMultiValueMap<String, String>().apply{ this.add("error", message) }
```
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply api/app/src/main/kotlin/packit/security/oauth2/OAuth2FailureHandler.kt:29]: The `UriComponentsBuilder` which the redirect utility uses takes a `MultiValueMap` (a map to a list) and `LinkedMultiValueMap` is the implementation of that interface.

[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply api/app/src/main/kotlin/packit/security/oauth2/OAuth2FailureHandler.kt:29]: Will rename to `queryString`!

**Comment by EmmaLRussell** [comment]
> LGTM and works like a charm... Just want to double check what org permissions are needed... eg. i had mrc-ide but have to request approval to be able to use it....i only could use reside as i am an owner

As we tested: looks like the org details permissions only have to be granted once to the app, by an org owner, not per user.

**Comment by EmmaLRussell** [comment]
> Are we going to handle the case where someone has the wrong secret for the id? Because currently you get this:

Oh, I wow hadn't seen that! :see_no_evil: 

> Are we also going to handle the wrong client id? because right now we go to a 404 page on github, wonder if we can just see the response from github and then tell the user their id does not exist? (these are probably suggestions for other tickets unless any of them are unintended)

Yeah, good idea, have added tickets for both of those. 
https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4750
https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4751

Hopefully these should be rare errors for real users to see once we've got an instance up and running, so maybe not high priority.

**Comment by EmmaLRussell** [comment]
> One last weird thing (but this might not be relevant, could be to do with how the app is brought up) if you:
    login with github successfully
    kill the docker containers running on your machine
    change the id in application properties
    re build everything
    on load you are immediately logged in and can do everything even though your client id is currently incorrect
Not sure if this is just because I have done something weird but perhaps we are not guarding certain routes correctly on load?

This is because the front end saves the JWT token to local storage and re-uses it on reload. As long as the token is still valid, the app will allow the user to login, and it won't go back to github to authenticate the user. So this is expected behaviour.

**Comment by M-Kusumgar** [comment]
> Yeah, good idea, have added tickets for both of those. https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4750 https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4751
> 
> Hopefully these should be rare errors for real users to see once we've got an instance up and running, so maybe not high priority.

Yep no worries! thanks!

**Review by absternator — APPROVED** [review]
LGTM

**Pull request merged**